pandas/core/generic.py

```diff
@@ -8988,26 +8992,40 @@ def _align_series(
                     other.index, how=join, level=level, return_indexers=True
                 )
 
-            left = self._reindex_indexer(join_index, lidx, copy)
+            if is_series:
+                left = self._reindex_indexer(join_index, lidx, copy)
+            elif join_index is None:
+                left = self.copy() if copy else self
+            else:
+                data = {
+                    c: algos.take_nd(
```

Comment:
is there a reason not to use the BlockManager method here?

Reply:
I am not sure what exactly you mean by BlockManager method. The way I came to my solution was that I tried to replicate what `Series._reindex_indexer` does. So what I am doing here is inspired by [this](https://github.com/pandas-dev/pandas/blob/main/pandas/core/series.py#L4520-L4523).

Reply:
> I am not sure what exactly you mean by BlockManager method.

I was thinking `Manager.take` (definitely only use public Manager methods)

> EDIT: The CI has already bitten me with an BlockManager ArrayManager issue (see for example [here](https://github.com/pandas-dev/pandas/runs/5300010405?check_suite_focus=true)).

yah, ArrayManager is a PITA.

Reply:
I tried 
```python
left = self.take(lidx)
left.index = join_index
```
but that makes `frame/test_align.py::test_align_series_combinations` and a lot of tests in `frame/test_arithmetic.py` fail. The issue is, that when `other.index` has elements which are not in `self.index`. Then `lidx` contains `-1` and so `.take()` will put in the last element instead of `np.nan`.

Consider the following example:
```python
import pandas as pd

df = pd.DataFrame({'x': [1., 2., 3., 44.], 'y': [5., 6., 7., 88.]}, index=pd.Index([1, 2, 8, 9]))
series = pd.Series([10., 20., 30., 40.], index=pd.Index([1, 2, 3, 4]))

join_index, lidx, rindx = df.index.join(series.index, how="outer", return_indexers=True)

with_algos_take_nd = pd.DataFrame(
    pd.core.algorithms.take_nd(df.values, lidx), columns=['x', 'y'], index=join_index
)
with_take = df.take(lidx)
with_take.index = join_index

print("with take_nd")
print(with_algos_take_nd)
print("with take")
print(with_take)
```
gives
```
with take_nd
      x     y
1   1.0   5.0
2   2.0   6.0
3   NaN   NaN
4   NaN   NaN
8   3.0   7.0
9  44.0  88.0
with take
      x     y
1   1.0   5.0
2   2.0   6.0
3  44.0  88.0
4  44.0  88.0
8   3.0   7.0
9  44.0  88.0
```

Reply:
> I tried [...] `left = self.take(lidx)`

The name overlap can be confusing: by "Manager.take` i meant `self._mgr.take`, which behaves differently from `self.take`.